aten/src/ATen/core/op_registration/op_registration.h

```diff
@@ -774,45 +786,55 @@ namespace detail {
 
 }
 
-// Represents a namespace in which we can define operators.  Conventionally
-// constructed using "torch::import".
+namespace detail {
+  class TorchLibraryInit;
+}
+
+// This is the "handle" by which functions defined in TORCH_LIBRARY
+// and TORCH_LIBRARY_IMPL can define operators and override implementations
+// at certain backends.
 //
-//      torch::import()
-//        .def("aten::add", ...)
-//        .def("aten::mul", ...)
+// Conventionally, you get access to it using those two macros:
 //
-class CAFFE2_API Module final {
-  c10::optional<std::string> ns_;
-
-  std::vector<RegistrationHandleRAII> registrars_;
-
-  Module(std::string ns);
-  Module();
-
-  // Use these as the constructors
-  friend Module _import_DOES_NOT_WORK_WITH_MOBILE_CUSTOM_BUILD(std::string ns);
-  friend Module import();
-
-private:
-  // Non-user visible actual implementations of functions.  These aren't
-  // public because we only implement & qualifier and not && qualifier
-  Module& _def(FunctionSchema&& schema) &;
-  Module& _def(c10::either<OperatorName, FunctionSchema>&&, CppFunction&& f) &;
-  Module& _impl(const char* name, CppFunction&& f) &;
-  Module& _fallback(CppFunction&& f) &;
-
+// TORCH_LIBRARY(torchvision, m) {
+//    // m is a c10::Library
+//    m.def("roi_align", ...);
+//    ...
+// }
+//
+// TORCH_LIBRARY_IMPL(aten, XLA, m) {
+//    // m is a c10::Library
+//    m.impl("add", ...);
+//    ...
+// }
+//
+// In some cases, you need to define something that applies to all namespaces,
+// not just one namespace (usually a fallback).  In that case, use the reserved
+// namespace _, e.g.,
+//
+// TORCH_LIBRARY_IMPL(_, XLA, m) {
+//    m.fallback(xla_fallback);
+// }
+//
+class CAFFE2_API Library final {
 public:
-  Module(const Module&) = delete;
-  Module& operator=(const Module&) = delete;
+  // Which type of macro produced this Library
+  enum Kind {
+    DEF, // from TORCH_LIBRARY (no qualifier)
+    IMPL,
+    FRAGMENT,
+  };
+
+  // Use TORCH_LIBRARY/TORCH_LIBRARY_IMPL instead of these constructors directly
```

Comment:
Note to myself: when Library constructor is used directly the current analyzer will fail.

Reply:
Only done in tests right now